Force recompilation if imported module has changed

hubertp opened the pull request:
### Pull Request Description

IR cache never really took into account a situation when a binding from the imported module has changed. In other words, it would continue to happily use the serialized metadata without noticing that it changed.

This change forces cache invalidation when any of the imported modules was invalidated (or rather not loaded from cache). 

### Important Notes

Added simple test infrastructure that simulates file modifications that would trigger the initial cache invalidation.
If they succeed, cache invalidation is propagated thus causing an error.

### Checklist

Please include the following checklist in your PR:

- [ ] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

hubertp committed:
Force recompilation if imported module has changed

IR cache never really took into account a situation when a binding from
the imported module has changed. In other words, it would continue to
happily use the serialized metadata without noticing that it changed.

This change forces cache invalidation when any of the imported modules
was invalidated (or rather not loaded from cache).
Sadly, it doesn't appear that we have appropriate infrastructure to
write such scenarios so I'm delaying that bit for now.

hubertp committed:
Update docs for IR caching

radeusgd reviewed (approved):
Looks good to me.

As we don't have tests, we need to ensure manually that it works as expected:
1. Did you try the scenario from the bug report to verify that now it behaves correctly?
2. Can you somehow verify if after the changes there are not "too" many invalidations? (I guess just running a program twice, one with cleared caches and one cached would be ok to see that the second one is faster - very imperfect test but better than nothing)

hubertp committed:
Test setup for checking cache invalidation logic

Using _x, where x is the change step, to simulate modifications to files
and thus trigger cache invalidation that should propagate to upstream
modules.

hubertp commented:
> Looks good to me.
> 
> As we don't have tests, we need to ensure manually that it works as expected:
> 
> 1. Did you try the scenario from the bug report to verify that now it behaves correctly?
> 2. Can you somehow verify if after the changes there are not "too" many invalidations? (I guess just running a program twice, one with cleared caches and one cached would be ok to see that the second one is faster - very imperfect test but better than nothing)

Yeah, I did all of that. 
In the end I figured it wasn't so hard to setup tests for this. This was added in the last commit

radeusgd commented on engine/runtime/src/test/scala/org/enso/compiler/test/ModifiedTest.scala:
```diff
@@ -0,0 +1,122 @@
+    copySources(from, to, filter)
+  }
+
+  private def copyIterationFiles(from: File, to: File, iteration: Int): Unit = {
```
This iteration logic is a bit complicated to me.

I understand that we match all files containing `_`? This seems possibly a bit too much. For now it works, but in case we wanted to re-use this class in the future - maybe we can add some special substring, like `Iterated` to the module name to mark it so? Or put it in a subdirectory `Iterations`/`Iterated` or something?

This is a nitpick, as it works good, but reading this code I needed to give it some think to see what is going on. Maybe at least add a comment to this function explaining in short what logic it is supposed to implement? (That it will select files containing `_` and select only the ith file from such sets.

hubertp replied:
Not entirely. It matches all files ending with `_<digit>+.enso` suffix which should limit it. Besides, this only applies to projects specified in the test suite. I would prefer not to have a separate directory (at least not for this case) as it becomes harder to manage the contents of the project. 
I added some comments to clarify the logic but I'd prefer not to spend more time on it. If new use-cases arise it will be rather simple to change this.

radeusgd replied:
Sure, it's already really cool. The comments are good, thank you :)

radeusgd reviewed (approved):
That's even better, great!

I really appreciate adding the tests, as not having tests really means we cannot guarantee we don't regress - see the REPL being broken so often.

hubertp committed:
Update CHANGELOG

hubertp committed:
Address PR comments

JaroslavTulach commented on engine/runtime/src/test/resources/Test_Caching_Invalidation/src/Main.enso:
```diff
@@ -0,0 +1,6 @@
+from project.Atom import all
+
+main =
+    false = 0 == 1
```
Why not just `IO.println "hmm..."`?

hubertp replied:
Symbol has to be used IIRC. This particular statement was because I wanted to avoid importing stdlib but in the end that was unavoidable. I can remove the `false` declaration though. Good point.

JaroslavTulach commented on engine/runtime/src/test/scala/org/enso/compiler/test/ModifiedTest.scala:
```diff
@@ -0,0 +1,134 @@
+import java.io.{ByteArrayOutputStream, File, FileFilter}
+import java.nio.file.{Files, Paths, StandardCopyOption}
+
+trait ModifiedTest
```
Excellent testing harness!

JaroslavTulach commented on engine/runtime/src/main/scala/org/enso/compiler/Compiler.scala:
```diff
@@ -218,7 +218,23 @@ class Compiler(
+      val modules = runImportsAndExportsResolution(module)
+      if (
+        module
+          .wasLoadedFromCache() && modules.exists(!_.wasLoadedFromCache())
```
Does the `modules.exists` check just direct dependencies of a module? What if there is a transitive dependency that is invalidated? Three modules depends one on another `A -> B ->C`. `C` caches are out of date. When `A` module is being checked it is OK and `B` is also OK. But then `B` is checked and gets invalidated because of `C`.

Will such situation be handled correctly?

hubertp replied:
Yes, it will. Thanks to the way the current import/export resolution is handled.
I double checked :)

JaroslavTulach commented on docs/runtime/ir-caching.md:
```diff
@@ -201,6 +202,19 @@ working state. This means that:
 
+### Imports
+
+Integrity Checking does not check the situation when the cached module imports a
```
Git, BitCoin, etc. would try to solve this situation by including hashes of all imported modules into the hash of each module to create a _(block) chain_. Couldn't such solution be used in this case as well? Why?

hubertp replied:
We could implement something like Merkle tree as well but it didn't seem necessary for the purpose of this fix.
Eventually, if we want to be more efficient in the kind of modules that we recompile we probably should combine hashes but I would say that this is a separate task.

JaroslavTulach reviewed (approved):
Great test! One question about transitive dependencies & ordering. One question about including hashing with a _chain_. Otherwise, looks like a progress. Approving.

hubertp committed:
PR review

mergify[bot] committed:
Merge branch 'develop' into wip/hubert/cache-invalidation-182953733

hubertp committed:
file copying is problematic for windows (and its 2022)
